accepted/0000-workspaces.md

```diff
@@ -0,0 +1,481 @@
+
+
+```
+$ tree
+.
+├── apps
+│  ├── x
+│  │  └── node_modules
+│  │     ├── bar -> ../../../node_modules/bar
+│  │     ├── baz -> ../../../node_modules/baz
+│  │     └── foo -> ../../../node_modules/foo
+│  ├── y
+│  │  └── node_modules
+│  │     ├── bar -> ../../../node_modules/bar
+│  │     ├── baz -> ../../../node_modules/baz
+│  │     └── foo -> ../../../node_modules/foo
+│  └── z
+│     └── node_modules
+│        ├── bar -> ../../../node_modules/bar
+│        ├── baz -> ../../../node_modules/baz
+│        └── foo -> ../../../node_modules/foo
+├── node_modules
+│  ├── bar -> ../packages/bar
+│  ├── baz -> ../packages/baz
+│  ├── foo -> ../packages/foo
+│  ├── x -> ../apps/x
+│  ├── y -> ../apps/y
+│  └── z -> ../apps/z
+├── package.json
+├── packages
+│  ├── bar
+│  │  └── node_modules
+│  │     ├── baz -> ../../../node_modules/baz
+│  │     └── foo -> ../../../node_modules/foo
+│  ├── baz
+│  │  └── node_modules
+│  │     ├── bar -> ../../../node_modules/bar
+│  │     └── foo -> ../../../node_modules/foo
+│  └── foo
+│     └── node_modules
+│        ├── bar -> ../../../node_modules/bar
+│        └── baz -> ../../../node_modules/baz
+├── plugins
+│  ├── a
+│  │  └── node_modules
+│  │     ├── baz -> ../../node_modules/baz
+│  │     └── foo -> ../../node_modules/foo
+│  ├── b
+│  │  └── node_modules
+│  │     ├── baz -> ../../node_modules/baz
+│  │     └── foo -> ../../node_modules/foo
+│  ├── c
+│  │  └── node_modules
+│  │     ├── baz -> ../../node_modules/baz
+│  │     └── foo -> ../../node_modules/foo
+│  ├── d
+│  │  └── node_modules
+│  │     ├── baz -> ../../node_modules/baz
+│  │     └── foo -> ../../node_modules/foo
+│  ├── node_modules
+│  │  ├── baz -> ../../node_modules/baz
+│  │  └── foo -> ../../node_modules/foo
+│  └── package.json
+└── ws2
+   ├── apps
+   │  └── twoapp
+   │     └── node_modules
+   │        └── two -> ../../../node_modules/two
+   ├── node_modules
+   │  ├── bar -> ../../node_modules/bar
+   │  ├── big-external-dep
+   │  ├── two -> ../packages/two
+   │  └── twoapp -> ../apps/twoapp
+   ├── package.json
+   └── packages
+      └── two
+         └── node_modules
+            ├── bar -> ../../../node_modules/bar
+            └── big-external-dep -> ../../../node_modules/big-external-dep
+```
+
+
+## Prior Art
+
+- [Lerna](https://github.com/lerna/lerna)
+- [Yarn workspaces](https://classic.yarnpkg.com/en/docs/workspaces/)
+- [Pnpm workspaces](https://pnpm.js.org/en/workspaces)
+
+## Unresolved Questions and Bikeshedding
+
+- Should we add a Workspace class (subclass of Node) in Arborist?
+- For this initial implementation there's no intention of adding a more ellaborate version/publish subcommand/workflow that would allow for bumping versions of nested packages + updating dependency references elsewhere across a workspace (similar to `lerna --independent` publish workflow).
+- Support for non-nested file structure workspaces (ex: adjacent directories)
```

Comment:
I'm curious what this would look like or if it's worth bringing into the discussion. From what I've seen, a common file structure is something like:

```
├── package.json
├── server
│  ├── package.json
├── front-end
│  ├── package.json
```
Usually there are scripts included at the root `package.json` and the dependencies are listed at the nested `package.json`s with their own scripts, such as install and build. Do you see the root `package.json` being the "main" directory? Also, do you think there will be any issues with this structure?

Reply:
Note: For the sake of making the entire conversation around **workspaces** more clear we went ahead and defined a **Dictionary** section that you can see [over here](https://github.com/npm/rfcs/blob/8a5080c881a771d25638d66e5150e4ea8e39c7a2/accepted/0000-workspaces.md#dictionary) - it defines the "main directory" as being the **Top-level workspace**. That said, any script in the **Top-level workspace** can be run as usual using `npm run` BUT it will still be necessary to use some sort of userland module such as [Lerna](https://lerna.js.org/) to help out with running `scripts` across nested **workspaces**.

With regards to the provided example, I think this is exactly the kind of structure we're aiming to support so def I see no issues 😊 as I mentioned in previous comments, the nested **workspaces** `run-scripts` support and other **npm commands** are going to be tackled in a follow up implementation defined at #117